presto-main/src/main/java/com/facebook/presto/execution/buffer/SpoolingOutputBuffer.java

```diff
@@ -0,0 +1,775 @@
+                totalRowsAdded.get(),
+                totalPagesAdded.get(),
+                ImmutableList.of());
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return state.get() == BufferState.FINISHED;
+    }
+
+    @Override
+    public double getUtilization()
+    {
+        return totalInMemoryBytes.get() / (double) threshold.toBytes();
+    }
+
+    @Override
+    public boolean isOverutilized()
+    {
+        return totalInMemoryBytes.get() > threshold.toBytes();
+    }
+
+    @Override
+    public long getPeakMemoryUsage()
+    {
+        return peakMemoryUsage.get();
+    }
+
+    @Override
+    public ListenableFuture<?> isFull()
+    {
+        return immediateFuture(null);
+    }
+
+    @Override
+    public void addStateChangeListener(StateMachine.StateChangeListener<BufferState> stateChangeListener)
+    {
+        state.addStateChangeListener(stateChangeListener);
+    }
+
+    @Override
+    public void setOutputBuffers(OutputBuffers newOutputBuffers)
+    {
+        requireNonNull(newOutputBuffers, "newOutputBuffers is null");
+        checkArgument(outputBuffers.getType() == SPOOLING, "Invalid output buffers type");
+        checkArgument(outputBuffers.isNoMoreBufferIds(), "invalid noMoreBufferIds");
+
+        if (state.get().isTerminal() || outputBuffers.getVersion() >= newOutputBuffers.getVersion()) {
+            return;
+        }
+
+        outputBuffers.checkValidTransition(newOutputBuffers);
+    }
+
+    @Override
+    public void enqueue(Lifespan lifespan, List<SerializedPage> pages)
+    {
+        if (!state.get().canAddPages()) {
+            return;
+        }
+        PendingRead pendingRead;
+        synchronized (this) {
+            this.pages.addAll(pages);
+
+            long bytesAdded = getPagesSize(pages);
+            long pagesAdded = pages.size();
+
+            // update output buffer info
+            totalBufferedBytes.addAndGet(bytesAdded);
+            totalBufferedPages.addAndGet(pagesAdded);
+            totalPagesAdded.addAndGet(pagesAdded);
+            totalRowsAdded.addAndGet(getPagesRows(pages));
+
+            totalInMemoryBytes.addAndGet(bytesAdded);
+
+            totalPagesRemaining.addAndGet(pagesAdded);
+            peakMemoryUsage.accumulateAndGet(totalInMemoryBytes.get(), Math::max);
+
+            if (totalInMemoryBytes.get() > threshold.toBytes()) {
+                flush();
+            }
+
+            pendingRead = this.pendingRead;
+            this.pendingRead = null;
+        }
+
+        if (pendingRead != null) {
+            processPendingRead(pendingRead);
+        }
+    }
+
+    @Override
+    public synchronized void enqueue(Lifespan lifespan, int partition, List<SerializedPage> pages)
+    {
+        checkState(partition == 0, "Expected partition number to be zero");
+        enqueue(lifespan, pages);
+    }
+
+    private synchronized void flush()
+    {
+        List<DataOutput> dataOutputs = pages.stream()
```

Comment:
Check drainTo(Collection<? super E> col) in BlockingQueue to take a snapshot if the queue. this will also empty the queue. So no need to do pages.clear(). After that, `synchronized` can be removed.